### DISCUSSION TITLE
Setting up a block diagonal (blocks separated by variables) preconditioner

### DISCUSSION THREAD

**Discussion by gridley** [idaholab/moose#17775] [Q&A General] [answered]
Hi all,

So, I have a Jacobian of the form:

![image](https://user-images.githubusercontent.com/18088906/117517105-f19d2780-af68-11eb-8594-bc2c9300b1b1.png)

Not very surprising, since you could write any matrix like this. The key is that the blocks are separated by the variables. The top row is a pressure Poisson equation, and the bottom row is a momentum equation. I know that A is a Poisson operator, so I'd like to use multigrid here. I'd like to apply block-diagonal preconditioning like:

![image](https://user-images.githubusercontent.com/18088906/117517167-1ee9d580-af69-11eb-8655-6e0ad56625d7.png)

Where \tilde A^{-1}, an approximation to the true inverse of A, comes from one or two applications of boomeramg or the like, and \tilde D^{-1}, an approximation to the inverse of D, would come from some other as-yet specified preconditioner I know will work well on that type of physics.

So, I think there's some way to do this with pcfieldsplit. Could any one point me at an example like this? It seems there is some undocumented Split.h file that may do this. Not sure which is the best way to approach this.

EDIT: after thinking on this some more, this request maybe doesn't make sense. If anyone recommends reading that explains theory behind the object in Split.h, I'd really like to hear.

Thanks for your time,

Gavin

**Comment by gridley** [comment, marked as answer]
To partially answer my own question, there are references to explain the different field split approaches in the MOOSE incompressible navier stokes module paper. I'm still not sure whether these are best applied using the Split object in Split.h or via Petsc arguments.

https://arxiv.org/abs/1710.08898

**Reply by aeslaughter** [reply]
@fdkong Can you help please?

**Reply by gridley** [reply]
Hey, I had messaged Fande in the MOOSE slack and think I figured it out. You have to use `-pc_type fieldsplit -pc_fieldsplit_type schur -pc_fieldsplit_schur_fact_type diag`. Then there are tons of other options to set up the field split solve, but this seems to converge as expected, albeit seeming to have some serious associated overhead with setting up intermediate matrices needed for this approach.
